File: server/src/utils/macroCalculator.ts
export function calculateBMR(
  weight_kg: number,
  height_cm: number,
  age: number,
  gender: string
): number {
  // Mifflin-St Jeor Equation
  if (gender.toLowerCase() === 'male') {
    return 10 * weight_kg + 6.25 * height_cm - 5 * age + 5;
  } else {
    return 10 * weight_kg + 6.25 * height_cm - 5 * age - 161;
  }
}

const activityMultipliers: Record<string, number> = {
  sedentary: 1.2,
  light: 1.375,
  moderate: 1.55,
  active: 1.725,
  very_active: 1.9,
};

export function calculateTDEE(bmr: number, activityLevel: string): number {
  const multiplier = activityMultipliers[activityLevel] ?? 1.2;
  return Math.round(bmr * multiplier);
}

export interface MacroTargets {
  calories: number;
  protein: number;
  carbs: number;
  fat: number;
}

export function calculateMacros(
  tdee: number,
  goal: string,
  weight_kg: number
): MacroTargets {
  let calories: number;

  switch (goal.toLowerCase()) {
    case 'cut':
      calories = tdee - 500;
      break;
    case 'bulk':
      calories = tdee + 300;
      break;
    case 'maintain':
    default:
      calories = tdee;
      break;
  }

  // Protein: ~2.2g per kg (~1g per lb)
  const protein = Math.round(2.2 * weight_kg);

  // Fat: 25% of total calories / 9 cal per gram
  const fat = Math.round((calories * 0.25) / 9);

  // Carbs: remainder / 4 cal per gram
  const proteinCalories = protein * 4;
  const fatCalories = fat * 9;
  const carbs = Math.round((calories - proteinCalories - fatCalories) / 4);

  return {
    calories: Math.round(calories),
    protein,
    carbs,
    fat,
  };
}

File: ios/Cairn/Utils/MacroCalculator.swift
import Foundation

struct MacroTargets {
    let calories: Double
    let protein: Double
    let carbs: Double
    let fat: Double
}

struct MacroCalculator {

    // MARK: - BMR (Mifflin-St Jeor)

    static func calculateBMR(weightKg: Double, heightCm: Double, age: Int, gender: String) -> Double {
        let base = 10 * weightKg + 6.25 * heightCm - 5 * Double(age)
        return gender.lowercased() == "male" ? base + 5 : base - 161
    }

    // MARK: - TDEE

    static func calculateTDEE(bmr: Double, activityLevel: String) -> Double {
        let multiplier: Double
        switch activityLevel.lowercased() {
        case "sedentary":   multiplier = 1.2
        case "light":       multiplier = 1.375
        case "moderate":    multiplier = 1.55
        case "active":      multiplier = 1.725
        case "very_active": multiplier = 1.9
        default:            multiplier = 1.55
        }
        return bmr * multiplier
    }

    // MARK: - Macro Targets

    static func calculateMacros(tdee: Double, goal: String, weightKg: Double) -> MacroTargets {
        let adjustedCalories: Double
        switch goal.lowercased() {
        case "cut":      adjustedCalories = tdee - 500
        case "bulk":     adjustedCalories = tdee + 300
        default:         adjustedCalories = tdee
        }

        let protein = 2.2 * weightKg
        let fat = (adjustedCalories * 0.25) / 9.0
        let proteinCals = protein * 4.0
        let fatCals = fat * 9.0
        let carbs = (adjustedCalories - proteinCals - fatCals) / 4.0

        return MacroTargets(
            calories: adjustedCalories,
            protein: protein,
            carbs: carbs,
            fat: fat
        )
    }

    // MARK: - Unit Conversions

    static func lbsToKg(_ lbs: Double) -> Double {
        lbs / 2.20462
    }

    static func kgToLbs(_ kg: Double) -> Double {
        kg * 2.20462
    }

    static func inchesToCm(_ inches: Double) -> Double {
        inches * 2.54
    }

    static func cmToInches(_ cm: Double) -> Double {
        cm / 2.54
    }
}
